### PR TITLE
README: document KCL technical decisions/params

### PR DESCRIPTION
**Pull request by arl**
#### :question: What

Document technical decisions, parameters and high-level implementation of the KCL Baker input.

#### :hammer: How to test

1. List all steps necessary;
2. To test this pull request.

#### :white_check_mark: Checklists

_This section contains a list of checklists for common uses, please delete the checklists that are useless for your current use case (or add another checklist if your use case isn't covered yet)._

- [ ] Is there unit/integration test coverage for all new and/or changed functionality added in this PR?
- [ ] Have the changes in this PR been functionally tested?
- [ ] Has `make gofmt-write` been run on the code?
- [ ] Has `make govet` been run on the code? Has the code been fixed accordingly to the output?
- [ ] Have the changes been added to the [CHANGELOG.md](./CHANGELOG.md) file?
- [ ] Have the steps in [CONTRIBUTING.md](./CONTRIBUTING.md) been followed to update a Go module?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -387,6 +391,62 @@ The [./examples/help/](./examples/help/) folder contains a working example of
...
+The dynamodb table also serves the purpose of checkpointing, that is keeping
+track of the per-shard advancement by writing the ID last read record
+(checkpoint).
+
+You can choose the initial position to read from when the application starts
+(that is when the dynamodb table doesn't exist) by setting InitialPosition
+either to LATEST or TRIM_HORIZON.
+
+
+##### Implementation and prevent throttling prevention
```
Comment (tommyblue) [inline README.md:428]: uhm, too many "prevent"?

File: README.md [hunk 2]
```diff
@@ -387,6 +391,62 @@ The [./examples/help/](./examples/help/) folder contains a working example of
...
+
+Leases are updated at regular interval defined by ShardSync.
+
+The dynamodb table also serves the purpose of checkpointing, that is keeping
+track of the per-shard advancement by writing the ID last read record
+(checkpoint).
+
+You can choose the initial position to read from when the application starts
+(that is when the dynamodb table doesn't exist) by setting InitialPosition
+either to LATEST or TRIM_HORIZON.
```
Comment (tommyblue) [inline README.md:425]: I'd add a suggestion for the value to use (rephrase as you like the most):


Note that those values are used only when a checkpoint doesn't exist in DynamoDB for the shard. If the checkpoint exists, its value is always used first.

##### How to choose the correct InitialPosition value 

When chosing the InitialPosition value, have in mind that `TRIM_HORIZON` can be useful when a resharding happens. In that case, in fact, before baker starts reading the new shards, some records could have already been published there. In this case using `LATEST` leads to have those records ignored. On the other hand, using `TRIM_HORIZON` means that, the first time that baker starts reading an existing stream, all the records hold in the shards must be read (data retention period depends on the stream configuration and is variable between 24 hours to 7 days). If reading all the existing records is a problem, then a possible strategy is to start baker the first time with `LATEST` (to create the first checkpoint) and then switch to `TRIM_HORIZON`

File: README.md [hunk 3]
```diff
@@ -387,6 +391,62 @@ The [./examples/help/](./examples/help/) folder contains a working example of
...
+
+##### Implementation and prevent throttling prevention
+
+Within a Baker instance, the KCL input creates as many record processors as
+there are shards to read from.  A record processor pulls records by way of the
+[GetRecords](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_GetRecords.html)
+AWS Kinesis API call.
+
+AWS [imposes limits on GetRecords](https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html), 
+Each shard can support up to a maximum total data read rate of 2 MiB per second
```
Comment (tommyblue) [inline README.md:436]:
```suggestion
each shard can support up to a maximum total data read rate of 2 MiB per second
```

File: README.md [hunk 4]
```diff
@@ -387,6 +391,62 @@ The [./examples/help/](./examples/help/) folder contains a working example of
...
+
+### Inputs
+
+#### KCL
+
+`input.KCL` fetches records from AWS [Kinesis](https://aws.amazon.com/kinesis/)
+using [vmware-go-kcl](https://github.com/vmware/vmware-go-kcl), an
+implementation of the KCL (Kinesis Client Library).
+KCL provides a way to to process a single Kinesis stream from multiple Baker
+instances, each instances reading a specific set of shards.
```
[the remark this answers is not in the document]
Comment (arl) [reply README.md:405]: I rephrased but didn't use _max_ since the default is 32767, in other words, there are not limits.

File: README.md (hunk 1 above)
Comment (arl) [reply README.md:428]: done

File: README.md [hunk 5]
```diff
@@ -387,6 +391,68 @@ The [./examples/help/](./examples/help/) folder contains a working example of
...
+Leases are updated at regular interval defined by ShardSync.
+
+The dynamodb table also serves the purpose of checkpointing, that is keeping
+track of the per-shard advancement by writing the ID last read record
+(checkpoint).
+
+InitialPosition defines the initial checkpoint position for consuming new
+shards. This parameter is only effective the first time a shard ID is
+encountered, since after that the lease will associate the shard and a record
+checkopint. It can either be set to LATEST or TRIM_HORIZON.
```
Comment (tommyblue) [inline README.md:426]:
```suggestion
checkpoint. It can either be set to LATEST or TRIM_HORIZON.
```

File: README.md [hunk 6]
```diff
@@ -387,6 +391,68 @@ The [./examples/help/](./examples/help/) folder contains a working example of
...
+
+Note that when new shards are created in the event of a resharding, KCL may not
+immediately be aware of their creation. Setting TRIM_HORIZON is thus a safer
+choice here since eventually all the records from the newly shards will be
```
Comment (tommyblue) [inline README.md:430]:
```suggestion
choice here since eventually all the records from the newly created shards will be
```

**Review by tommyblue — APPROVED** [review]
lgtm, fixed 2 typos

**Pull request merged**